### PR TITLE
Builder trinary minmax

### PR DESCRIPTION
**Pull request by trenouf**
This adds only one new builder method, for fmed3, implemented as the
fmed3 instruction for float, or a combination of fmin and fmax for half.

The SPIR-V reader has been changed to use that builder method, and for
all other trinary_minmax instructions it just generates IR itself using
standard LLVM IR instructions and intrinsics.

Change-Id: If0f3d9e54c6ff79fc1a13871f817bb9eeccb0465

### CODE REVIEW

**Comment by trenouf** [comment]
Please review just the topmost commit.

**Comment by trenouf** [comment]
V3: Can use fmed3 for half on gfx9+.

File: builder/llpcBuilderImplArith.cpp [hunk 1]
```diff
@@ -36,3 +36,769 @@
...
+{
+    Value* pIsPositive = CreateFCmpOGT(pInValue, Constant::getNullValue(pInValue->getType()));
+    Value* pPartialResult = CreateSelect(pIsPositive, ConstantFP::get(pInValue->getType(), 1.0), pInValue);
+    Value* pIsNonNegative = CreateFCmpOGE(pPartialResult, Constant::getNullValue(pInValue->getType()));
+    return CreateSelect(pIsNonNegative, pPartialResult, ConstantFP::get(pInValue->getType(), -1.0), instName);
+}
+
+// =====================================================================================================================
+// Create derivative calculation on float or vector of float or half
+Value* BuilderImplArith::CreateDerivative(
```
Comment (linqun) [inline builder/llpcBuilderImplArith.cpp]: please move it to misc group.

File: builder/llpcBuilderImplArith.cpp [hunk 2]
```diff
@@ -637,6 +637,44 @@ Value* BuilderImplArith::CreateDerivative(
...
 
+// =====================================================================================================================
+// Create "fmed3" operation, returning the middle one of three scalar or vector float or half values.
+Value* BuilderImplArith::CreateFMed3(
+    Value*        pValue1,    // [in] First value
+    Value*        pValue2,    // [in] Second value
+    Value*        pValue3,    // [in] Third value
+    const Twine&  instName)   // [in] Name to give instruction(s)
+{
+    // For float, and for half on gfx9+, we can use the fmed3 instruction. We need to scalarize this ourselves.
```
Comment (amdrexu) [inline builder/llpcBuilderImplArith.cpp]: gfx9 -> GFX8

File: builder/llpcBuilderImplArith.cpp [hunk 3]
```diff
@@ -637,6 +637,44 @@ Value* BuilderImplArith::CreateDerivative(
...
+                                      return CreateIntrinsic(Intrinsic::amdgcn_fmed3,
+                                                             pValue1->getType(),
+                                                             { pValue1, pValue2, pValue3 });
+                                   });
+        pResult->setName(instName);
+        return pResult;
+    }
+
+    // For half on gfx8 or earlier, use a combination of fmin and fmax.
+    FastMathFlags fmf;
```
Comment (amdrexu) [inline builder/llpcBuilderImplArith.cpp]: fmf -> fastMathFlags

File: builder/llpcBuilderImplArith.cpp [hunk 4]
```diff
@@ -637,6 +637,44 @@ Value* BuilderImplArith::CreateDerivative(
...
+                                   {
+                                      return CreateIntrinsic(Intrinsic::amdgcn_fmed3,
+                                                             pValue1->getType(),
+                                                             { pValue1, pValue2, pValue3 });
+                                   });
+        pResult->setName(instName);
+        return pResult;
+    }
+
+    // For half on gfx8 or earlier, use a combination of fmin and fmax.
```
Comment (amdrexu) [inline builder/llpcBuilderImplArith.cpp]: gfx8 -> GFX8

File: builder/llpcBuilderImplArith.cpp (hunk 1 above)
Comment (trenouf) [reply builder/llpcBuilderImplArith.cpp]: I am resolving this in #157.

**Comment by trenouf** [comment]
The resolution of the review issues will arrive when I push the rebased version of this change after #157 is merged.

**Comment by trenouf** [comment]
V4: Rebased and addressed review comments.

This is now a single commit.

**Comment by amdvlk-admin** [comment]
### Test summary for commit b9328a646832919bfc0cbcd01c0abc87363b44bb
<details open><summary>Driver commits used in build</summary><ul>

<li>METROHASH: amd-mainline 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 68b57dba33a4d922e8f1ef1b3781c2f659ffbd1c</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 331558e93794068a786bf699d3fe23bb11bac021</li>
<li>LLVM: amd-vulkan-dev f0a8af5e93d57b890687c7ced43943a0c3a167ca</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 925/925 (100.0%)</li>
<li>Failed: 0/925 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 366250/366250 (100.0%)</li>
<li>Failed: 0/366250 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 282963/282963 (100.0%)</li>
<li>Failed: 0/282963 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**